### DISCUSSION TITLE
XTLS-Reality (or any other sing-box/xray-core config before Mullvad Wireguard server)

### DISCUSSION THREAD

**Discussion by imagecreation1** [mullvad/mullvadvpn-app#9119] [Feature requests & Ideas]
### I have checked if others have suggested this already

- [X] I have checked this issue tracker to see if others have reported similar issues.

### Feature description

Hello. Thanks for the great VPN service!
I just wanted to ask you to add support for sing-box https://github.com/SagerNet/sing-box (or xray-core) configs to bypass hardcore censorship. Now XTLS-Reality is heavily used by people in countries like China or Iran to bypass censorship. Currently you're using old things like vmess/vless in your application without the uTLS fingerprint library and other features. 
It would be great if we could add our XTLS-Reality in the chain before Mullvad's Wireguard server.

### Alternative solutions

You can use sing-box's https://sing-box.sagernet.org/configuration/outbound/wireguard/#structure wireguard outbound with detour (https://sing-box.sagernet.org/configuration/shared/dial/#detour). Or maybe we can manually exclude sing-box.exe using Split Tunneling and use a local proxy, which it will publish as inbound, to connect to Mullvad's Wireguard server.

### Type of feature

- [ ] Better privacy/anonymity
- [X] Better at circumventing censorship
- [ ] Easier to use
- [ ] Other

### Operating System

- [ ] Android
- [ ] iOS
- [X] Windows
- [ ] macOS
- [X] Linux

**Comment by rslnrdnk** [comment]
I second this. I was Mullvad user in Russia before, but after Russian state blocked OpenVPN & WG servers, I was forced to move on to other, proxy provider. I'd gladly return back to Mullvad if they'd implement this.

**Comment by bebanowski** [comment]
May be necessary in the UK soon if the government decides to ban VPNs that are increasingly being used to circumvent the recent adult content block.

**Comment by curable-online** [comment]
> use a local proxy, which it will publish as inbound, to connect to Mullvad's Wireguard server.

How do you do this?  I have local xray/sing-box; I don't know how tunnel Mullvad connection through it.

**Comment by testerfeedy** [comment]
Adding +1 from Russia, August 2026.

Real-world situation:
- ISP: Beeline (mobile), TSPU deployed
- All obfuscation dead: QUIC, Shadowsocks 443, UDP-over-TCP 443, LWO
- Result: completely unable to use paid subscription on mobile

Current workaround: None for Android/iOS. Desktop XTLS guide works, but requires technical skills most users don't have.

This is not a "nice to have" feature anymore. For users in Russia, China, Iran — this is the difference between "can use Mullvad" and "cannot use Mullvad at all."

Please consider reopening as issue or adding to roadmap. Even SOCKS5 proxy support (smaller scope than full sing-box) would solve 90% of cases.

cc: anyone from Russia/China/Iran — please comment if you're affected.

**Comment by LindaFerum** [comment]
@testerfeedy I think Mullvad team could reduce the amount of effort on their side by simply (okay "simply") enabling configuration of shadowsocks UDP config since many VLESS+ clients can open a shadowsocks inbound and modern android has reasonable localhost security (so Mullvad pumping traffic to 127.0.0.1:9999 unencrypted shadowsocks UDP would be a good way to wrap it in  VLESS/XTLS/Reality via say Karing)
